lib/uncruft/deprecate_attribute.rb

```diff
@@ -0,0 +1,30 @@
+module Uncruft
+  module DeprecateAttribute
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      def deprecate_attribute(attribute, aliased_attribute: nil, message:)
```

Comment:
I think it would be really cool if we made `message` optional and auto-generated a method based on the inputs.

Basically, if you don't specify an alias, we output a simple method on read or on write:
- `[attribute] is deprecated. Please avoid using it.`
- `[attribute]= is deprecated. Please avoid using it.`

And if you do specify an alias, we can say:
- `[attribute] is deprecated. Please use [aliased_attribute] instead.`
- `[attribute]= is deprecated. Please use [aliased_attribute]= instead.`

Thoughts?

Reply:
I'm concerned that the risk of someone not realizing they can provide a custom message when they should outweighs the benefit of saving yourself a minute or so of crafting something simple.

From the examples where I would have liked to use this, I don't think an auto-generated message would capture everything I want to convey.

- `Please avoid using it` is probably not enough context in a lot of situations. I would imagine in most cases where we're deprecating something, there's a preferred alternative or some sort of explanation for how to reframe what we're trying to accomplish.
- Sometimes we split a single concept into two or more separate concepts, so the alias may always be the correct alternative. For example, if we had a `name` attribute and wanted to split it into `first_name` and `nickname`, we don't want to say `Please  use first_name`. We want to say something like `Please use first_name in formal contexts and nickname in casual contexts.`

Do you share those concerns? If we want to provide some default option, maybe you could pass `message: :default`?

Reply:
Yeah `message: :default` could work. I mean I could go either way on this. I do agree that in a lot of cases it makes sense to specify a custom message. However, it's also possible to use `ActiveSupport::Deprecation.warn` directly without much additional effort beyond what this mixin provides. That sort of led me to think that we want to focus on more targeted use cases, like "attribute X has been renamed to Y", etc, and the more targeted the use case, the more natural it feels to auto-generate a message.

Basically, since we're intentionally constraining the problem, we're inherently starting to classify the ways in which a method might be deprecated, so in my mind I'd expect us to converge on similar messages that correspond to each of those classifications.

Reply:
Yeah, fair enough. If we see a strong majority of cases that fall within the same patterns, I'll be a lot more convinced. With the goal of shipping this feature quickly and learning more about how folks decide to use it, maybe we should bookmark this as v2 idea to revisit?

Reply:
Agreed 👍